ir.ExternKernel: correctly handle kwarg default arguments

benjaminglass1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #144124
* #144123
* #144002
* #143909
* #143421
* #143223
* __->__ #141371

Additionally, enable torchinductor opinfo tests exercising all
previously fixed bugs in this stack.

Note: I've manually sharded the cpp_wrapper CI checks into 2 shards.
Once all OpInfo tests are enabled we should switch back to automatic
sharding, but until then the pipeline doesn't have appropriate timing
stats.  More shards would be helpful given the compilation slowdown
associated with cpp_wrapper, but 2 will do for now.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update on "ir.ExternKernel: correctly handle kwarg default arguments"




[ghstack-poisoned]

benjaminglass1 committed:
Update on "ir.ExternKernel: correctly handle kwarg default arguments"




[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 commented:
@desertfire I just thought of an implementation bug in this PR. Hang on, I'm going to convert it back to draft while I fix it quickly.

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update on "ir.ExternKernel: correctly handle kwarg default arguments"


Properly query default arguments for kwargs, rather than always assuming `None`. Fixes a bug in fallback ops for `cpp_wrapper`.

[ghstack-poisoned]

benjaminglass1 committed:
Update on "ir.ExternKernel: correctly handle kwarg default arguments"


Properly query default arguments for kwargs, rather than always assuming `None`. Fixes a bug in fallback ops for `cpp_wrapper`.

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

desertfire commented on torch/_inductor/ir.py:
```diff
@@ -6503,9 +6508,16 @@ def export_extern_kernel_node(self):  # type: ignore[no-untyped-def]
+
+        # Supply any needed kwarg default values.  Any args with default values are
+        # already correctly set thanks to fill_non_provided_args.
+        kwarg_to_value = {
```
Does `get_kwargs_value` work here?

benjaminglass1 replied:
lol I didn't notice that function. Thanks for the pointer; I'll swap it in!

desertfire reviewed (commented):
Let's start to include some specific unit test from inductor/test_torchinductor_opinfo instead of waiting for https://github.com/pytorch/pytorch/pull/141699 to mature. You can also include relevant unit tests from your previous PRs on this stack.

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 commented:
@desertfire I squashed in the next commit in the stack that enabled `OpInfo` testing, and limited the tests down to things exercising most of the bugs I've fixed. Can you quickly re-review?

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 committed:
Update

[ghstack-poisoned]

desertfire commented on .ci/pytorch/test.sh:
```diff
@@ -382,15 +382,28 @@ test_inductor_aoti() {
-  # Run certain inductor unit tests with cpp wrapper. In the end state, we should be able to run all the inductor
-  # unit tests with cpp wrapper.
-  python test/run_test.py --include inductor/test_torchinductor.py --verbose
+  if [[ "$1" -eq "2" ]]; then
```
I don't see an early return or an else-branch. Won't this cause shard 2 to run both opinfo tests and all other tests?

benjaminglass1 replied:
Gah, you're so right. I'll fix that.

benjaminglass1 committed:
Update

[ghstack-poisoned]

benjaminglass1 commented:
@pytorchbot merge

benjaminglass1 committed:
Update

[ghstack-poisoned]
